### PR TITLE
Generic config_tree implementation

### PR DESCRIPTION
**Pull request by Miauwkeru**
- Add basic unix registry parts
- Create Config parser
- simplify get logic
- Detect configuration files
- add todo
- Add a generic space delimited config parser
- Update hacked together "registry"

### CODE REVIEW

File: dissect/target/plugins/os/unix/config_tree.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+                new_values = old_value + [new_values]
+            new_info[key] = new_values
+
+        self.parsed_data = new_info
+
+
+CONFIG_MAP: dict[str, LinuxConfigurationParser] = {
+    "ini": Ini,
+}
+
```
Comment (cecinestpasunepipe) [inline dissect/target/plugins/os/unix/config_tree.py:99]:
```suggestion
KNOWN_FILES: dict[str, LinuxConfigurationParser] = {
    "ulogd.conf": Ini,
}
```

Comment (cecinestpasunepipe) [reply dissect/target/plugins/os/unix/config_tree.py:99]: I think there are 2 base cases: extensions and well known files.

File: dissect/target/plugins/os/unix/config_tree.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    def __init__(self, fs: Filesystem, path: str, entry: Any, parser_items=None, collapse=None) -> None:
+        super().__init__(fs, path, entry)
+        if parser_items is None:
+            self.parser_items = self.parse_config(entry, collapse)
+        else:
+            self.parser_items = parser_items
+
+    def parse_config(self, entry: FilesystemEntry, collapse=None) -> ConfigParser:
+        extension = entry.path.rsplit(".", 1)[-1]
+        parser = CONFIG_MAP.get(extension, Unknown)(collapse)
```
Comment (cecinestpasunepipe) [inline dissect/target/plugins/os/unix/config_tree.py:202]:
```suggestion
        parser = CONFIG_MAP.get(extension, KNOWN_FILES.get(entry.name, Unknown))(collapse)
```

File: dissect/target/plugins/os/unix/config_tree.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+
+    def _write_value_mapping(self, output: io.BytesIO, data: dict[str, Any]):
+        if isinstance(data, list):
+            for x in data:
+                output.write(bytes(x, "utf-8"))
+                output.write(b"\n")
+        elif hasattr(data, "keys"):
+            output.write(b"\n")
+            for key, value in data.items():
+                output.write(bytes(key, "utf8"))
```
Comment (cecinestpasunepipe) [inline dissect/target/plugins/os/unix/config_tree.py:227]:
```suggestion
                output.write(bytes(key, "utf-8"))
```

File: dissect/target/plugins/os/unix/config_tree.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+            # Currently trying to open the underlying entry
+            return self.entry.open()
+
+        if self.is_dir():
+            bytesio = io.BytesIO()
+            self._write_value_mapping(bytesio, self.parser_items)
+            return bytesio
+        return io.BytesIO(bytes(self.parser_items, "utf8"))
+
+    def iterdir(self):
```
Comment (cecinestpasunepipe) [inline dissect/target/plugins/os/unix/config_tree.py:248]: return type?

File: dissect/target/plugins/os/unix/config_tree.py [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+            output.write(b"\n")
+            for key, value in data.items():
+                output.write(bytes(key, "utf8"))
+                self._write_value_mapping(output, value)
+        else:
+            output.write(b" ")
+            output.write(bytes(data))
+            output.write(b"\n")
+
+    def open(self):
```
Comment (cecinestpasunepipe) [inline dissect/target/plugins/os/unix/config_tree.py:234]: return type?

File: dissect/target/plugins/os/unix/config_tree.py [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+            bytesio = io.BytesIO()
+            self._write_value_mapping(bytesio, self.parser_items)
+            return bytesio
+        return io.BytesIO(bytes(self.parser_items, "utf8"))
+
+    def iterdir(self):
+        for entry in self.scandir():
+            yield entry.name
+
+    def scandir(self):
```
Comment (cecinestpasunepipe) [inline dissect/target/plugins/os/unix/config_tree.py:252]: return type?

File: dissect/target/plugins/os/unix/config_tree.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+                parser.read_file(open_file)
+        return parser
+
+    def get(self, path: str) -> ConfigurationEntry:
+        # Check for path in config entry
+        if path in self.parser_items:
+            return ConfigurationEntry(self.fs, path, self.entry, self.parser_items[path])
+        raise NotADirectoryError(f"Cannot open a {path!r} on a value")
+
+    def _write_value_mapping(self, output: io.BytesIO, data: dict[str, Any]):
```
Comment (cecinestpasunepipe) [inline dissect/target/plugins/os/unix/config_tree.py:219]: return type?

**Review by cecinestpasunepipe — CHANGES_REQUESTED** [review]

File: dissect/target/plugins/os/unix/config.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+            return
+        except MissingSectionHeaderError:
+            pass
+
+        fh.seek(offset)
+        open_file = io.StringIO("[DEFAULT]\n" + fh.read())
+        self.parsed_data.read_file(open_file)
+
+
+class Unknown(LinuxConfigurationParser):
```
Comment (cecinestpasunepipe) [inline dissect/target/plugins/os/unix/config.py:84]: Is this really unknown? It seems to me this is the standard Unix config style. I would recommend to name this Default instead.

**Review by cecinestpasunepipe — CHANGES_REQUESTED** [review]

**Review by cecinestpasunepipe — APPROVED** [review]

**Review by Schamper — COMMENTED** [review]
I also wonder if aligning closer with the Windows registry implementation would make more sense, instead of the filesystem implementation.

**Pull request merged**